### DISCUSSION TITLE
fatal error: opencv2/dnn/dnn.hpp: No such file or directory

### DISCUSSION THREAD

**Discussion by andregirard37** [OE4T/meta-tegra#1276] [Q&A] [answered]
Hello,

I am building a Yocto image with Dunfell but adding in my custom layer the opencv_4.5.5 recipe as well as the bbappend for CUDA dnn, that you could find on the Kirkstone release.
After building the rootfs, I believe I have cudnn available. However, for development purposes, I require the sysroot I am generating to have the dnn header files (ex: opencv2/dnn/dnn.hpp), for cross compiling applications on my host laptop. I tried adding as many opencv* recipes I could but unsuccessful. I added:
opencv
opencv-samples
libopencv-core-dev
libopencv-highgui-dev
libopencv-imgproc-dev
libopencv-ml-dev
Any idea what recipe I need to add that would include all the dnn header files ?
Thanks!

**Comment by dwalkes** [comment]
>  However, for development purposes, I require the sysroot I am generating to have the dnn header files (ex: opencv2/dnn/dnn.hpp), for cross compiling applications on my host laptop.

Are you using the [Extensible SDK](https://docs.yoctoproject.org/3.1.25/sdk-manual/sdk-extensible.html?highlight=sdk) or [Standard SDK](https://docs.yoctoproject.org/3.1.25/sdk-manual/sdk-using.html) for your host laptop development?  I would expect header files to end up in the SDK but not in the target rootfs.  I suspect there's probably not a recipe for including development header files on the target rootfs today.

**Reply by andregirard37** [reply]
Thanks for your reply. I don't require my rootfs to have the header files, but the sysroot to have them, so that I can cross compile not on the target but on my host laptop.
To build the SDK, I am running:  bitbake <image> -c populate_sdk 
Not sure if that builds the extensible SDK or the standard one to be honest.
Thanks

**Reply by dwalkes** [reply]
> I don't require my rootfs to have the header files, but the sysroot to have them, so that I can cross compile not on the target but on my host laptop.

OK understood.

> To build the SDK, I am running: bitbake  -c populate_sdk
> Not sure if that builds the extensible SDK or the standard one to be honest.

I think that would be the [standard sdk](https://docs.yoctoproject.org/3.1.25/sdk-manual/sdk-appendix-customizing-standard.html#adding-individual-packages-to-the-standard-sdk).

> but adding in my custom layer the opencv_4.5.5 recipe as well as the bbappend for CUDA dnn, that you could find on the Kirkstone release.

Can you reproduce this using [tegra-demo-distro](https://github.com/OE4T/tegra-demo-distro), one of the demo images there, and one of the kirkstone branches?   Something like `bitbake -c populate_sdk demo-image-full`? I'm trying to understand if this is a problem with meta-tegra kirkstone or your custom layer.  If this is something I can reproduce with tegra-demo-distro I will try to reproduce it.

**Reply by andregirard37** [reply]
Thanks for the reply. I was not aware of the tegra-demo-distro repo. Looking at the github page, it seems that openCV would be indeed included on the demo-image-full image. So I will try that one and let you what I find. Thanks!

**Reply by andregirard37** [reply]
Alright, so that demo-image-full build has the /opencv4/opencv2/dnn folder. Which is progress. The problem is that this is using kirkstone but in my case I would want to stay with Dunfell. I'll dig more into what I could be missing. Again, I only pruned the opencv 4.5.5 recipe from Kirkstone, but anything else is Dunfell. Thanks again.

**Reply by dwalkes** [reply]
> that demo-image-full build has the /opencv4/opencv2/dnn folde
> The problem is that this is using kirkstone but in my case I would want to stay with Dunfell

And what happens if you use https://github.com/OE4T/tegra-demo-distro/tree/dunfell ?

**Reply by andregirard37** [reply]
I tried and I see the same behaviour as my own build. ../opencv4/opencv2/dnn does not exist.
Also, we want to use opencv_4.5.1 or greater and not opencv_4.1.0, as starting in 4.5.1, opencv adds dnn support for CUDA, which I need.
SO what I thought I could do is to build a Dunfell image, but in my custom layer adding the opencv 4.5.5 recipes (bb and bbappend). I would hope this is possible, however it appears I am missing something to have ../opencv4/opencv2/dnn in the sysroot.
Thank you.

**Reply by madisongh** [reply]
Does your bbappend contain the lines in [our kirkstone bbappend](https://github.com/OE4T/meta-tegra/blob/2a808b6afb221974196d2eeab3537140ebb63cc7/external/openembedded-layer/recipes-support/opencv/opencv_4.5.%25.bbappend), including applying the patch for locating CuDNN?  If you look at the `log.do_configure` output for your opencv 4.5.5 build, does it show it trying to detect whether `cudnn` is available?

**Reply by andregirard37** [reply]
Sorry your link does not work. However, in my setup, I included the opencv_4.5.5.bb, the opencv_4.5.%.bbappend. I also had to create the class/ folder to add the missing setuptools3-base.bbclass on Dunfell. And in the opencv sub folder I have:
0001-3rdparty-ippicv-Use-pre-downloaded-ipp.patch
0001-Temporarliy-work-around-deprecated-ffmpeg-RAW-functi.patch
0001-core-vsx-update-vec_absd-workaround-condition.patch
0001-Use-Os-to-compile-tinyxml2.cpp.patch
0001-Dont-use-isystem.patch
0003-To-fix-errors-as-following.patch
0001-Fix-search-paths-in-FindCUDNN.cmake.patch 
download.patch
0001-Make-ts-module-external.patch
I also have the ade recipe and subfolder in the main opencv folder. Not sure what this is. For now I just copied it over from kirkstone.
Thanks

**Reply by madisongh** [reply]
Sorry, I've fixed the link.  Now verify that the the Cmake configuration is actually finding CuDNN in the recipe sysroot, by checking the `log.do_configure` output.   If it isn't, make sure that the `0001-Fix-search-paths-in-FindCUDNN.cmake.patch` is getting applied and that you see the CuDNN files under `recipe-sysroot` in the opencv build tree.  If the patch isn't getting applied, you maybe missed adding it to the SRC_URI variable; if the CuDNN files aren't in the recipe sysroot, you maybe missed adding `cudnn` to DEPENDS (or to the `PACKAGECONFIG[cuda]` dependency list).

**Reply by andregirard37** [reply]
I am looking at build/tmp/work/aarch64-poky-linux. But I can't see opencv there, or cudnn/dnn. 
I can find opencv here: build/tmp/work/armv8a_tegra194-poky-linux/opencv/4.5.5-r0/temp
But I don't see log.do_configure. I have:

log.do_packagedata_setscene
run.buildhistory_emit_outputsigs.3573579
run.do_package_qa_setscene.3573578
run.sstate_unpack_package.3573578
log.do_packagedata_setscene.3573994
run.buildhistory_emit_outputsigs.3573580  
run.do_package_write_rpm_setscene          
run.sstate_unpack_package.3573579
log.do_package_qa_setscene                 
run.buildhistory_emit_outputsigs.3573633  
run.do_package_write_rpm_setscene.3573579  
run.sstate_unpack_package.3573580
log.do_package_qa_setscene.3573578         
run.buildhistory_emit_outputsigs.3573994  
run.do_populate_lic_setscene               
run.sstate_unpack_package.3573633
log.do_package_write_rpm_setscene          
run.buildhistory_emit_pkghistory.3573578  
run.do_populate_lic_setscene.3573580       
run.sstate_unpack_package.3573994
log.do_package_write_rpm_setscene.3573579 
run.buildhistory_emit_pkghistory.3573579  
run.do_populate_sysroot_setscene           
run.toaster_packagedata_dumpdata.3573994
log.do_populate_lic_setscene               
run.buildhistory_emit_pkghistory.3573580  
run.do_populate_sysroot_setscene.3573633   
run.uninative_changeinterp.3573578
log.do_populate_lic_setscene.3573580 
run.buildhistory_emit_pkghistory.3573633  
run.do_rm_work                             
run.uninative_changeinterp.3573579
log.do_populate_sysroot_setscene           
run.buildhistory_emit_pkghistory.3573994  
run.do_rm_work.3620614                     
run.uninative_changeinterp.3573580
log.do_populate_sysroot_setscene.3573633   
run.buildhistory_emit_sysroot.3573633     
run.sstate_hardcode_path_unpack.3573578    
run.uninative_changeinterp.3573633
log.do_rm_work                            
run.buildhistory_list_pkg_files.3573994   
run.sstate_hardcode_path_unpack.3573579    
run.uninative_changeinterp.3573994
log.do_rm_work.3620614                     
run.do_packagedata_setscene               
run.sstate_hardcode_path_unpack.3573580
log.task_order                            
run.do_packagedata_setscene.3573994       
run.sstate_hardcode_path_unpack.3573633
run.buildhistory_emit_outputsigs.3573578  
run.do_package_qa_setscene                
run.sstate_hardcode_path_unpack.3573994

I checked the bbappend file and since it is the exact copy of the kirkstone one, I think the SRC_URI is correct:
SRC_URI:append:cuda = " file://0001-Fix-search-paths-in-FindCUDNN.cmake.patch"
As well:
PACKAGECONFIG[cuda] = "${@opencv_cuda_flags(d)},-DWITH_CUDA=OFF,${CUDA_DEPENDS} cudnn"

**Reply by madisongh** [reply]
It's under `armv8a_tegra194-poky-linux` because the CUDA modifications make it SoC-specific.  Looks like a prior build is getting pulled in via the shared state cache.  Try `bitbake -c clean -c cleansstate opencv`, then `bitbake opencv` again to rebuild from the beginning.

**Reply by andregirard37** [reply]
Thanks! In the log.do_configure I see:
-- Module opencv_objdetect disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_barcode disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dnn_objdetect disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dnn_superres disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dpm disabled because opencv_objdetect dependency can't be resolved!
-- Module opencv_face disabled because opencv_objdetect dependency can't be resolved!
-- Module opencv_mcc disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_wechat_qrcode disabled because opencv_dnn dependency can't be resolved!

Not sure if that is a red flag or not, but regarding the patch, I see the following in log.do_patch.3798776:
NOTE: Applying patch '0001-Fix-search-paths-in-FindCUDNN.cmake.patch' (../layers/meta-custom/recipes-support/opencv/opencv/0001-Fix-search-paths-in-FindCUDNN.cmake.patch)
DEBUG: Python function patch_do_patch finished
DEBUG: Python function do_patch finished
DEBUG: Executing python function do_qa_patch
DEBUG: Python function do_qa_patch finished

I just tried regenerating the sysroot just in case, but still no dnn folder there.

Just to make sure... in my sysroots I only have:
sysroots/aarch64-poky-linux and sysroots/x86_64-pokysdk-linux
Thanks

**Reply by madisongh** [reply]
The sysroot I'm talking about is for the recipe during the build, which would be under `tmp/work..../opencv/<version>/recipe-sysroot`.

It should be fairly obvious in the `log.do_configure` messages whether it's found or not.  If it is, you'll see a line near the top of that log that looks like this:
```
-- Found CUDNN: /build/madison/tegra-demo-32.7.x/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.5.5-r0/recipe-sysroot/usr/lib/libcudnn.so (found suitable version "8.2.1", minimum required is "7.5") 
```
and in the summary at the bottom of the log you would see:
```
--   cuDNN:                         YES (ver 8.2.1)
```

If all that's there, but you don't see `dnn`,  `dnn_objdetect`, and `dnn_superres` in the `To be built:` list of modules in that summary section, look above in the log for `opencv_dnn` to see if configuring that module generated some errors, or a message like `Module opencv_dnn disabled because <some reason>`.

**Reply by andregirard37** [reply]
Yes, the lines above in my comment are from the log.do_configure in  build/tmp/work/armv8a_tegra194-poky-linux/opencv/4.5.5-r0/temp/log.do_configure.
I don't see any reference to cuDNN or libcudnn. I only find, as mentioned above:
-- Module opencv_objdetect disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_barcode disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dnn_objdetect disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dnn_superres disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_dpm disabled because opencv_objdetect dependency can't be resolved!
-- Module opencv_face disabled because opencv_objdetect dependency can't be resolved!
-- Module opencv_mcc disabled because opencv_dnn dependency can't be resolved!
-- Module opencv_wechat_qrcode disabled because opencv_dnn dependency can't be resolved!

Which is what you are suggesting at the end of your last comment.

**Reply by dwalkes** [reply]
@andregirard37 can you fork tegra-demo-distro, add your changes in a branch and send a link?

**Reply by madisongh** [reply]
Or just post the recipe and bbappend(s) you're trying to use...

**Reply by andregirard37** [reply]
Here are the files (renamed txt to send...). Thanks
[opencv_4.5.%.txt](https://github.com/OE4T/meta-tegra/files/11821784/opencv_4.5.txt)
[opencv_4.5.5.txt](https://github.com/OE4T/meta-tegra/files/11821785/opencv_4.5.5.txt)

**Reply by andregirard37** [reply]
I am a bit confused on the following line in the recipe
PACKAGECONFIG[dnn] = "-DBUILD_opencv_dnn=ON -DPROTOBUF_UPDATE_FILES=ON -DBUILD_PROTOBUF=OFF,-DBUILD_opencv_dnn=OFF,protobuf protobuf-native,"
why is there DBUILD_opencv_dnn ON then OFF ? Is that to list the 2 available options ?
And also in the bbappend:
-DWITH_CUDA=OFF 
Shouldn't that be ON
Thanks

**Reply by andregirard37** [reply]
in to log.do_configure I see:
--   OpenCV modules:
--     To be built:                 alphamat aruco bgsegm bioinspired calib3d ccalib core cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaoptflow cudastereo cudawarping cudev datasets features2d flann fuzzy gapi hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor ml optflow phase_unwrapping photo plot python3 quality rapid reg rgbd saliency sfm shape stereo stitching structured_light superres surface_matching tracking ts video videoio videostab xfeatures2d ximgproc xphoto
--     **Disabled:                    dnn** freetype text world
**--     Disabled by dependency:      barcode cudaobjdetect dnn_objdetect dnn_superres** dpm face mcc objdetect wechat_qrcode xobjdetect
--     Unavailable:                 cvv hdf java julia matlab ovis python2 viz
--     Applications:                perf_tests examples apps
--     Documentation:               NO
--     Non-free algorithms:         NO

So there might be one recipe/dependency that is preventing dnn to build correctly ?

**Reply by madisongh** [reply]
[Yocto documentation on PACKAGECONFIG](https://docs.yoctoproject.org/3.1.25/singleindex.html#term-PACKAGECONFIG)

The `Disabled:` line means that OpenCV's `dnn` module was disabled in the configuration.  So something must not be right with the back-port of your bbappend, which should be enabling `cuda` and `dnn` in the configuration automatically for the Tegra platforms.

**Reply by dwalkes** [reply, marked as answer]
I think the problem is that the `TEGRA_CUDA_ARCHITECTURE` check [here](https://github.com/OE4T/meta-tegra/blob/kirkstone/external/openembedded-layer/recipes-support/opencv/opencv_4.5.%25.bbappend#L5-L9) doesn't work when copying to the dunfell branch because the commit which added this at https://github.com/OE4T/meta-tegra/commit/4a89e2bdcc8052bbf7b7dff4b5caeb8aa1903995 isn't used in the [dunfell tegra-194.inc](https://github.com/OE4T/meta-tegra/blob/dunfell/conf/machine/include/tegra194.inc).

I've verified if I add `TEGRA_CUDA_ARCHITECTURE = "72"` to tegra194.inc I get:
```
-- Found CUDNN: /home/dan/trellis/tegra-demo-distro/build-dunfell/tmp/work/armv8a_tegra194-oe4t-linux/opencv/4.5.5-r0/recipe-sysroot/usr/lib/libcudnn.so (found suitable version "8.2.1", minimum required is "7.5")
```
and
```
--   cuDNN:                         YES (ver 8.2.1)
```
in `log.do_configure`.

**Reply by andregirard37** [reply]
Thanks a lot! I added in my local.conf :
```bitbake
TEGRA_CUDA_ARCHITECTURE = "72"
``` 
and now I see the same lines:
```bitbake
-- Found CUDNN: /home/yocto/projects/l4w-nx/build/tmp/work/armv8a_tegra194-poky-linux/opencv/4.5.5-r0/recipe-sysroot/usr/lib/libcudnn.so (found suitable version "8.2.1", minimum required is "7.5")
``` 
And:
```bitbake
--   cuDNN:                         YES (ver 8.2.1)
``` 
Now I see the dnn/ folder in my sysroot! Thanks for all the support. Amazing!
